File: yaa/Tools/Finish.py
from yaa.Tools.BaseTool import Tool

class Tool(Tool):
    '''完成会话
    大模型使用时：
        大模型的回答包含：<完成会话>\n<理由>（对任务完成的总结）</理由>\n</完成会话>
    '''
    ToolInfo = {
        "id": "finish",
        "name": "完成会话",
        "description": "如果所有的任务和子任务都已经完成，使用这个工具结束会话。",
        "parameters": {
            "properties": [
                {
                    "name": "理由",
                    "type": "string",
                    "description": "对任务完成的总结"
                }
            ],
            "required": ["理由"]
        },
        "feedback": {
            "properties": [
                {
                    "name": "授权",
                    "type": "bool",
                    "description": "用户是否批准本次工具调用"
                },
                {
                    "name": "反馈",
                    "type": "string",
                    "description": "用户对自己决策的说明"
                }
            ],
            "required": ["授权"]
        }
    }

    @classmethod
    def use(cls, session_data):
        if session_data['config']['tool'][cls.ToolInfo['id']]['auto_approve'] == True:
            session_data['messages'].append({
                "role": "tool",
                "content": f"[完成会话]执行成功。"
            })
            session_data['status'] = '已完成'
        else:
            session_data['status'] = '已中断'
        return session_data

File: src/agent/tools/finish.rs
use crate::core::session::SessionData;
use crate::core::tool::{Tool, ToolError, ToolInput, ToolOutput, ToolParam};
use async_trait::async_trait;
use serde_json::json;

pub struct FinishTool;

#[async_trait]
impl Tool for FinishTool {
    fn name(&self) -> &str {
        "finish"
    }

    fn description(&self) -> &str {
        "当任务完成时调用此工具结束会话。需要提供任务完成的总结理由。"
    }

    fn parameters(&self) -> Vec<ToolParam> {
        vec![ToolParam {
            name: "reason".to_string(),
            description: "任务完成的总结理由".to_string(),
            required: true,
            r#type: "string".to_string(),
        }]
    }

    fn default_config(&self) -> crate::core::session::ToolApprovalConfig {
        crate::core::session::ToolApprovalConfig { auto_approve: true }
    }

    async fn execute(
        &self,
        input: ToolInput,
        session: &mut SessionData,
    ) -> Result<ToolOutput, ToolError> {
        // 检查自动授权配置
        if !session
            .config
            .tool
            .tools
            .get("finish")
            .map(|config| config.auto_approve)
            .unwrap_or(true)
        {
            return Err(ToolError::NotAuthorized);
        }

        // 修改会话状态为已完成
        session.status = crate::core::session::SessionStatus::Completed;

        // 添加工具调用消息到会话
        session.add_message(
            crate::core::session::Role::System,
            format!(
                "[{}]执行成功。任务完成理由：{}",
                self.name(),
                input
                    .param_values
                    .get("reason")
                    .and_then(|v| v.as_str())
                    .unwrap_or("未提供理由")
            ),
        );

        Ok(ToolOutput {
            name: self.name().to_string(),
            result: json!({"status": "completed"}),
        })
    }
}
